packet/src/pacer.rs

```diff
@@ -0,0 +1,131 @@
+use std::time::{Duration, Instant};
+
+pub struct LeakyBucketPacer {
+    /// Target bitrate in bits per second.
+    target_bitrate: u64,
```

Comment:
You may want to be aware that eventually you will actually want 3 rates, not just 1.

1.  The target rate: the rate given by the congestion controller that basically means "don't sent more than this"
2.  The padding rate: a value less than or equal to the target rate.  You would not send padding more than this, but you would send non-padding more than this.  Because sometimes you have to send padding, but you don't want to send padding beyond what would be useful for the application.
3.  Some rate related to draining the queue when it gets too big.  For libwebrtc's pacer, I think they can have this be larger than the target rate.  For the pacer I wrote, I instead allocated a portion of the target rate for draining and then had the target rate used elsewhere lowered.

Here is the libwebrtc code:

https://source.chromium.org/chromium/chromium/src/+/main:third_party/webrtc/modules/pacing/pacing_controller.h;l=223;drc=c14f4e9399fc4e23a6c4c68fc60086bdceec1c90;bpv=1;bpt=1

Reply:
Interesting, this is based somewhat on Pion's implementation which the two alternate rates you mentioned.

Padding seems easy enough to add later.

I've been assuming that the system that pushes the data into the pacer will respect the current bitrate here. My thinking is that if the input rate exceeds the target rate we should drop the oldest packets in the queue(s), this is what happens currently. It sounds like you are saying that we should attempt to burst them out instead. 

I think the answer to this might have significant implications on the design

Reply:
You definitely can *not* assume the rate coming into the pacer will be respected over short periods.  Traffic can be bursty.  And that's kind of the whole point of the pacer.

It's not really a good idea to just drop burst traffic.  It's often that the burst is really important.  It's better to queue and the figure out a better behavior, such as allocating part of the target send rate to draining the queue, or drain out higher priority packets before lower priority packets, or maybe drop the lower priority packets if the queue gets really too big. 

It can get somewhat complicated.  

And it doesn't need to be all done in this PR.  I just wanted to let you know kind of direction you might want to go down eventually.

Reply:
Awesome, this insight is super valuable.

I guess the pressing question is whether the design as outlined here will be problematic.

```
                                                                                                                            
                                        Actual buffers                                                                      
                                           in here                                                                          
                                                                                                                            
┌────────────────────────┐        ┌────────────────────────┐                                                                
│      Media::write      │───────▶│   PacketizingBuffer    ├─────────┐                                                      
└────────────────────────┘        └────────────────────────┘         │                                                      
                                                                     │          ┌────────────────────────┐                  
                                                                     ├─────────▶│         Pacer          │────poll_packet──▶
                                                                     │          └────────────────────────┘                  
┌────────────────────────┐        ┌────────────────────────┐         │                                                      
│      Media::write      │───────▶│   PacketizingBuffer    │─────────┘                 Store packets                        
└────────────────────────┘        └────────────────────────┘                          until next poll                       
                                                                                                                            
                                                                                                                            
                                        Actual buffers                                                                      
                                           in here                                                                          
                                                                                                                            ```

If choice around prioritisation need to be made I can see how this isn't great 🤔

Reply:
If choices around prioritisation need to be made I can see how this isn't great 🤔

Reply:
If I recall correctly, libwebrtc actually has multiple buffers that it round-robins through.  But that's done more as an implementation detail of the pacer.  I suppose it just depends on where you draw the line of where the "pacer" is.   You could view the "PacketizingBuffers+Pacer" in your diagram as the "whole Pacer" and the "Pacer" in your diagram as just one part of the "whole Pacer".

At the end of the day, regardless of what you call the pieces, you need:

1.  Something that cares a lot about time and decides _when_ to send a packet (controls when to leak).
2.  A place (or places) to store packets that are waiting for their turn to go (the bucket, or buckets).
3.  Something to decide which packet comes out when it's time to send one (controls what to leak)
4.  Something to decide what happens when the bucket(s) is (are) getting too full.
5.  A way to get or create padding packets when there is nothing stored (what to send when the bucket is empty).

It sounds like you'll do (5) later.

Your diagram seems like it could work for (1) and (2).  I guess you just need to determine how (3) and (4) happen.

Reply:
I was just looking at that code in libWebRTC that does prioritization. Seems easy enough to add later.

I think I'm gonna aim for 1, 2, and 4 initially, leaving 3 and 5 for later. For 4 this is what you were hinting at with a drain bitrate, right? 

For example, if the target bitrate is ~1Mbit/s(625 bytes per 5ms interval) and we find at the next send interval that we actually have (625 + **X** bytes) to send, we need to allow sending at least some, but maybe not all of **X**?

For #5, if we find that we have `400 bytes` we should make up the remaining 200 bytes with padding packets

Reply:
I don't think we need 2 in the first iteration.

Reply:
We definitely need it, but it doesn't need to be within the pacer. For now we can keep using the `PacketizingBuffer`s for it